[1단계 - 자동차 경주 구현] 에덴(김성산) 미션 제출합니다.

leo0842 opened the pull request:
안녕하세요 핀! 이렇게 만나뵙게 되어 영광입니다!

처음으로 페어 프로그래밍을 하면서 프리코스때 같은 미션을 할 때와는 다르게 TDD 방식으로 코딩도 해보고, 페어가 작성하는 코드들을 보면서 레이싱 미션 1차 PR을 하게 됐습니다. 다만 진행을 하면서 궁금했던 점이 있어 해당 방식을 먼저 진행하였던 핀에게 도움을 요청드리고자 합니다!

1. 커밋 단위와 커밋의 크기
- 페어와 커밋에 대해 이야기를 할 때, 기능 구현 단위로 커밋을 하자고 하였고, 리드미에 적은 기능 구현 목록마다 커밋을 하여 10개 미만의 커밋으로 1차 PR을 날리게 되었는데요, 다른 크루들의 PR 내용을 보니 커밋이 저희의 단위와는 차원이 다르게 커밋의 수가 많더라구요. 커밋의 단위도 잘게 되어 있었구요. 혹시 페어와 제가 진행한 커밋의 단위가 너무 큰 단위로 적은 수의 커밋만 한 것이 아닌지, 커밋에 대해 개선할 방안은 없는지 여쭤보고 싶습니다!

2. 앞으로의 리팩토링 방안
- 먼저 페어와 기본적인 기능 구현만 하여 PR을 날리게 되었는데, 앞으로 Iuput에 대한 예외 처리(이름에 특수 문자 또는 빈 값이 들어오는 경우) 말고 코드를 더 개선할 수 있는 리팩토링 방안이 있는지 여쭤보고 싶습니다!

긴 글 읽어주셔서 감사드리고, 좋은 주말 보내세요!!

leo0842 committed:
feat: TDD 문자열 덧셈 계산기 구현

leo0842 committed:
docs: 기능 구현 목록 추가

leo0842 committed:
feat: 이름을 입력받고 입력받은 차 이름에 대해 검증하는 로직 구현

leo0842 committed:
feat: 레이싱 횟수 입력 받는 로직 구현

leo0842 committed:
feat: 랜덤 값을 생성하고 전진 여부 판단 구현

leo0842 committed:
feat: 레이싱 진행 상황 출력 기능 구현

leo0842 committed:
feat: 우승자 판별 및 출력 기능 구현

leo0842 committed:
refactor: 문자열 상수화 및 메소드 추출 리팩토링

lalize commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,24 @@
+        this.location.increase();
+    }
+
+    public Integer getLocation() {
```
getLocation 네이밍과 달리 Location을 반환하는 게 아니네요.
네이밍을 바꿔보면 어떨까요?

```suggestion
    public Integer position() {
```

leo0842 replied:
지적 감사드립니다. getCurrentPosition() 으로 변경하였습니다.

lalize commented on src/main/java/racingcar/controller/Racing.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    private void showRacingResult() {
+        Referee referee = new Referee();
```
매번 Referee를 생성안해도 되지 않을까요?

lalize commented on src/main/java/racingcar/controller/Racing.java:
```diff
@@ -0,0 +1,50 @@
+
+    private void showRacingResult() {
+        Referee referee = new Referee();
+        board.recordRacingWinners(referee.judgeRacingWinners(participants.getParticipantCars()));
```
getter를 사용하지 말고 Participants에게 우승자를 구해달라고 하면 되지 않을까요?
Referee가 꼭 있어야 하나요?

leo0842 replied:
해당 부분에 대해 고민을 많이 해보았습니다. 
Participants 에게 많은 행동을 부여하는 것은 아닌지, 
Participants 내부에서 Referee 를 부르는 것은 어떤지, 
Referee 를 없애고 Participants 와 Car 끼리 상호작용 하는 것은 어떤지 등에 대해 여러 방법으로 생각해 보았습니다.
결론적으로는 말씀해주신 방법대로 Referee 를 없애는 방법으로 진행하였습니다. 굳이 Referee 라는 객체 없이도 우승자를 구하는 로직이 Participants 에게 과도한 책임도 아닐뿐더러 오히려 명확해질 수 있는 방법이라고 생각하여 Referee 객체를 없애는 방안으로 진행하게 되었습니다. 지적 감사드립니다!!

lalize commented on src/main/java/racingcar/model/Participants.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public void race() {
+        cars.stream().filter(car -> isMovable()).forEach(Car::move);
```
자동차 이동 판단이 Participants의 책임이 맞을까요?
Car 객체를 단독으로 사용하면 조건 없이 전진할 수 있을 것 같네요 🤔
추가로 자동차마다 이동 판단 로직이 다르다면? 어떻게 개선할 수 있을까요?

leo0842 replied:
1. 마찬가지로 코멘트해주신 내용이 적절한 것 같아 이동 판단에 대해서는 Car 가 판단하도록 하였고, Participants 는 Random 한 수를 생성하는 것까지로 책임을 제한하였습니다.
2. 자동차마다 이동 판단 로직이 다를 수 있다는 것에 대해 처음 생각해보게 되었습니다! 해당 부분은 막연히 생각나는 interface 를 이용하여 만들어 보게 되었습니다. tryMove() 를 참가하는 차 마다 override 하여 각자의 이동 판단 로직을 만들 수 있게 하였고, Participants 에서는 tryMove() 라는 행동만 요청하는 것으로 바꾸어 보았습니다.
이 부분에 대해 interface 를 작성한 것이 과도한 구조는 아닌지, 조금 더 나은 방향이 있는지에 대해 여쭤보고 싶습니다!

lalize replied:
interface를 사용하는 부분은 좋은 것 같아요.
그런데 interface를 Car가 구현하지 말고 Car가 interface를 의존 주입 받는 건 어떨까요?

lalize commented on src/main/java/racingcar/model/Participants.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public List<Car> getParticipantCars() {
+        return cars;
```
반환된 List<Car>를 외부에서 조작하면 Participants 객체가 영향을 받겠네요. 어떻게 방지할 수 있을까요?

leo0842 replied:
미처 외부에서 조작할 수 있다는 생각을 해보지 못하였습니다. 그래서 해당 부분에 대해서는 return 값으로 Collections 의 unmodifiablelist 메소드를 이용하여 수정할 수 없는 리스트를 반환하도록 하였고, 테스트도 추가하였습니다.

lalize commented on src/main/java/racingcar/model/Referee.java:
```diff
@@ -0,0 +1,25 @@
+            }
+            if (maxLocation == participantCar.getLocation()) {
+                winners.add(participantCar.getName());
+            }
```
> 규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.

leo0842 replied:
더욱 조심하겠습니다!

lalize commented on src/main/java/racingcar/model/Referee.java:
```diff
@@ -0,0 +1,25 @@
+            if (maxLocation < participantCar.getLocation()) {
+                winners = new ArrayList<>();
+                winners.add(participantCar.getName());
+                maxLocation = participantCar.getLocation();
```
max 값과 winners 를 계속해서 수정하고 있는데 전체에 대한 max 값을 먼저 구한 뒤 winners 를 구하면 더 읽기 좋은 코드가 되지 않을까요?

leo0842 replied:
읽기 좋은 코드에 대해 조금 더 주의깊게 생각하면서 로직을 작성하도록 노력하겠습니다. indent 부분과 엮어서 max 값을 구하고 winners 를 구하는 로직이 적절한 것 같아서 해당 방식으로 수정하였습니다.

lalize commented on src/main/java/racingcar/view/Staff.java:
```diff
@@ -0,0 +1,20 @@
+
+    private static final String INPUT_NAME_MESSAGE = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
+    private static final String INPUT_RACING_TIMES_MESSAGE = "시도할 회수는 몇회인가요?";
+    private static final Scanner scanner = new Scanner(System.in);
```
상수는 대문자로 바꿔주세요.

lalize commented on src/main/java/racingcar/view/Staff.java:
```diff
@@ -0,0 +1,20 @@
+        System.out.println(INPUT_RACING_TIMES_MESSAGE);
+        return Integer.parseInt(scanner.nextLine());
+    }
+}
```
사용자 입력을 받는 로직에 맞게 메서드 네이밍을 바꿔보는 게 어떨까요?

leo0842 replied:
이 부분도 '읽기 좋은 코드' 에 유념하여 작성하겠습니다! inputFromUser() 라는 메소드 명으로 바꾸었습니다.

lalize commented on src/test/java/racingcar/model/ParticipantsTest.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Test
+    public void race() {
```
race도 테스트 작성해주세요!

leo0842 replied:
race() 를 테스트하는 부분에 대해 여러 가지 시도를 해보았지만 결국 답을 찾지 못했습니다. race() 는 랜덤 값을 생성하는 private 메소드를 한 번 호출하고, car 에게 tryMove 라는 메세지를 보내는 역할을 가지고 있는데, 제가 시도해본 목록으로는
1. private 을 public 으로 바꾸어 랜덤 값을 강제로 부여하는 방법
  - 이 방법은 지양해야 하기 때문에 일찍 접었습니다.
2. private 에서 사용되는 Math.random() 을 mockito 의 when().thenReturn() 으로 조작하는 방법
  - Math 가 final 클래스여서 mock 으로 전환이 되지 않아 실패하였습니다. Math 클래스의 생성자가 private 이기때문에 인스턴스화도 불가능하였습니다. [관련 글](https://www.baeldung.com/mockito-final)
3. car에게 tryMove 메세지를 요청하는지만 알아보기 위해 verify() 확인
  - 역시 car 객체를 mock 으로 전환할 수 없어 실패하였습니다.

혹시 더 시도해 볼 수 있는 방법이 있는지 여쭤보고 싶습니다!

lalize commented on src/main/java/racingcar/model/Participants.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private int makeRandomNumberBetweenZeroAndNine() {
+        return (int) (Math.random() * 10);
```
10은 매직넘버가 아닌가요? 상수화하면 좋을 것 같아요!

leo0842 replied:
미처 발견하지 못하였습니다. 수정하였습니다!

lalize commented on src/main/java/racingcar/view/Staff.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.Scanner;
+
+public class Staff {
```
View의 책임을 갖고 있는 것 같은데 이름만 보면 View라고 파악하기 어렵네요.

leo0842 replied:
해당 부분에 대해서 가장 많은 시간을 쓴 것 같습니다.
Staff 라는 객체가 View 라는 책임이 표현되어야 하는지, 
본인의 역할에 맞는 네이밍으로만 작성하면 되는 것이 아닌지,
굳이 Staff 라는 이름이 필요한지에 대해 생각을 많이 해보게 되었습니다.
결론은 Staff 라는 객체의 본인의 역할이 Staff 라는 네이밍에서 완전하게 표현되지도 못하고 View 라는 책임도 표현하지 못하는 것 같아 InputView 라는 네이밍으로 바꾸게 되었습니다. 이에 따라 기존 OutputView 책임을 가지고 있던 Board 객체도 통일성을 맞추고자 OutputView 라는 네이밍으로 수정하였습니다.
지적 감사드립니다!

lalize commented on src/main/java/racingcar/model/Participants.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Participants {
```
일급 컬렉션 👍

leo0842 replied:
칭찬은 에덴을 춤추게 합니다 ㅎㅎ

lalize commented on src/main/java/racingcar/model/PlayTime.java:
```diff
@@ -0,0 +1,24 @@
+        return playTime == 0;
+    }
+
+    public void decreasePlayTime() {
```
뒤에 PlayTime을 붙이지 않아도 값 객체라서 의도가 잘 드러날 것 같아요!
```suggestion
    public void decrease() {
```

leo0842 replied:
값 객체에 대해 많이 서툴러서 해당 메소드명을 사용하게 되었습니다. 지적 감사드립니다!!

lalize commented on src/main/java/racingcar/model/PlayTime.java:
```diff
@@ -0,0 +1,24 @@
+        this.playTime = playTime;
+    }
+
+    public boolean isZero() {
```
isZero 보단 isEnd가 가독성 측면에서 나은 선택이지 않을까요?

leo0842 replied:
행동보다는 로직에 집중하여 isZero 라는 표현이 나온 것 같습니다. 코멘트 해주신대로 isEnd 가 적절해보여 수정하였습니다!

lalize reviewed (changes requested):
안녕하세요! 에덴🌻
객체의 역할과 책임에 대한 고민의 흔적이 보이네요~
전체적으로 잘 구현해주셨습니다👍
몇 가지 코멘트 남겼으니 확인 부탁드릴게요!

> 커밋 단위와 커밋의 크기
페어와 커밋에 대해 이야기를 할 때, 기능 구현 단위로 커밋을 하자고 하였고, 리드미에 적은 기능 구현 목록마다 커밋을 하여 10개 미만의 커밋으로 1차 PR을 날리게 되었는데요, 다른 크루들의 PR 내용을 보니 커밋이 저희의 단위와는 차원이 다르게 커밋의 수가 많더라구요. 커밋의 단위도 잘게 되어 있었구요. 혹시 페어와 제가 진행한 커밋의 단위가 너무 큰 단위로 적은 수의 커밋만 한 것이 아닌지, 커밋에 대해 개선할 방안은 없는지 여쭤보고 싶습니다!

`feat: 이름을 입력받고 입력받은 차 이름에 대해 검증하는 로직 구현` 커밋 변경사항의 경우 이름과 관련 없어 보이는 자동차의 위치 정보에 대한 코드, 여러 대의 자동차 만들어 레이싱에 참가시키는 로직들이 포함되어 있는 데 이런 부분들도 작게 나눠서 커밋해보시는 건 어떨까요?

> 앞으로의 리팩토링 방안

코멘트 확인 부탁드릴게요 ㅎㅎ

leo0842 committed:
refactor: getLocation 네이밍을 직관적으로 변경
getLocation -> getCarPosition 으로 변경

leo0842 committed:
refactor: 움직일 수 있는 행위에 대해 책임 변경
Participants -> Car 로 책임을 변경

leo0842 committed:
refactor: 네이밍과 파라미터 수정
move -> tryMove, 0개의 파라미터에서 1개의 파라미터로 수정

leo0842 committed:
test: 움직이지 말아야 하는 수를 받았을 때의 테스트

leo0842 committed:
refactor: 레이싱에 참가하는 차가 가져야할 메소드를 가진 인터페이스 추가
자동차마다 이동 판단 로직이 다를 수 있으므로 자동차는 RacingCar 인터페이스를 implements 하여 필요한 이동 판단 로직을 override 할 수 있도록 한다.

leo0842 committed:
refactor: 매 테스트시 인스턴스를 생성할 수 있는 setUp 메소드 추가

leo0842 committed:
refactor: getParticipantCars 호출 시 반환되는 List가 변경될 수 없도록 수정

leo0842 committed:
test: getParticipantCars 메소드에 대해 변경 불가 테스트 추가

leo0842 committed:
refactor: Referee 클래스의 역할을 Participants 로 변경

leo0842 committed:
refactor: 우승자의 위치를 묻는 메소드를 Car 에게 요청

leo0842 committed:
refactor: 상수 변수명 대문자로 변경

leo0842 committed:
refactor: 사용자로부터 입력을 받는 로직 메소드 명을 직관적으로 변경

leo0842 committed:
refactor: 1의 자리로 만들어주는 10 상수화

leo0842 committed:
refactor: View 의 역할을 담아내지 못한 Staff, Board 클래스 명을 변경

leo0842 committed:
refactor: 의도가 이미 드러나는 단어의 중복을 제거
PlayTime 이라는 객체로 이미 만들었기때문에 decreasePlayTime 메소드 이름보다는 decrease 가 적절

leo0842 committed:
refactor: isZero 를 조금 더 직관적인 isEnd 로 메소드명 변경

leo0842 committed:
test: participants 클래스의 우승자 찾는 로직 테스트

leo0842 committed:
refactor: BeforeEach 테스트로 중복되는 인스턴스 생성 제거

leo0842 committed:
test: Car 클래스의 우승자 위치의 차인지 확인하는 로직 테스트 추가

lalize commented on src/main/java/racingcar/model/Participants.java:
```diff
@@ -0,0 +1,54 @@
+public class Participants {
+    private static final int MAKE_ONE_DIGIT_MULTIPLIER = 10;
+
+    private final List<Car> cars;
```
final을 붙여준 이유를 말씀해주실 수 있으실까요?

lalize commented on src/main/java/racingcar/model/Location.java:
```diff
@@ -0,0 +1,14 @@
+package racingcar.model;
+
+public class Location {
```
원자 타입 포장 👍

lalize commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,39 @@
+    public Car(String carName) {
+        this.name = new Name(carName);
+        this.location = new Location();
+        this.movableNumber = 4;
```
현재 구조에서는 인스턴스 변수로 갖기 보단 상수화하는 게 더 적절해보여요.

lalize commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    public Integer getCarPosition() {
+        return location.getCurrentPosition();
+    }
+
+    public Name getName() {
+        return this.name;
+    }
```
getter 간 추상화 수준이 다른 것 같네요.

lalize commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,39 @@
+        return number >= movableNumber;
+    }
+
+    public boolean isWinnerPosition(Integer position) {
```
isWinnerPosition이 정말로 우승자 위치인지 확인하는 메서드가 맞나요?
단순히 입력으로 들어온 position 과 비교하는 메서드가 아닐까요?

lalize commented on src/main/java/racingcar/model/Participants.java:
```diff
@@ -0,0 +1,54 @@
+        this.cars.add(car);
+    }
+
+    public Integer participateSize() {
```
int가 아닌 Integer를 반환하는 이유가 궁급해요 :)

lalize commented on src/main/java/racingcar/model/Participants.java:
```diff
@@ -0,0 +1,54 @@
+        return Collections.unmodifiableList(cars);
+    }
+
+    public List<Name> findRacingWinners() {
+        Integer maxPosition = findMaxPosition();
+        return  cars.stream()
+                .filter(car -> car.isWinnerPosition(maxPosition))
+                .map(Car::getName)
+                .collect(Collectors.toList());
+    }
```
로직이 매우 심플하게 변경되었네요  👍

lalize commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,24 @@
+    private static final String INPUT_RACING_TIMES_MESSAGE = "시도할 회수는 몇회인가요?";
+    private static final Scanner SCANNER = new Scanner(System.in);
+
+    public String getCarNames() {
+        System.out.println(INPUT_NAME_MESSAGE);
+        return inputFromUser();
+    }
+
+    public Integer getPlayTimes() {
```
사용자 입력에 맞게 getXXX 보다는 다른 네이밍이 어떨까요?

lalize commented on src/test/java/racingcar/model/PlayTimeTest.java:
```diff
@@ -0,0 +1,14 @@
+
+import org.junit.jupiter.api.Test;
+
+class PlayTimeTest {
```
PlayTime 테스트가 부족한 것 같아요 :)

lalize commented on src/test/java/racingcar/model/ParticipantsTest.java:
```diff
@@ -0,0 +1,42 @@
+
+class ParticipantsTest {
+
+    private Participants participants;
+    @BeforeEach
```
```suggestion
    private Participants participants;
    
    @BeforeEach
```

lalize commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -0,0 +1,38 @@
+
+class CarTest {
+
+    private Car car;
+    @BeforeEach
```
```suggestion
    private Car car;
    
    @BeforeEach
```

lalize commented on src/main/java/racingcar/model/Name.java:
```diff
@@ -0,0 +1,28 @@
+public class Name {
+
+    private static final int MAX_NAME_LENGTH = 5;
+    private static final String MAX_NAME_LENGTH_ERROR_MESSAGE = "[ERROR] 이름은 5자 이하여야 합니다. ";
+    private final String name;
```
상수와 필드 사이에 빈줄을 추가하면 가독성이 더  좋아질 것 같아요.
```suggestion
    private static final String MAX_NAME_LENGTH_ERROR_MESSAGE = "[ERROR] 이름은 5자 이하여야 합니다. ";
    
    private final String name;
```

lalize commented on src/main/java/racingcar/model/Location.java:
```diff
@@ -0,0 +1,14 @@
+
+    private int location;
+
+    public int getCurrentPosition() {
```
```suggestion
    public int getLocation() {
```

lalize commented on src/main/java/racingcar/model/PlayTime.java:
```diff
@@ -0,0 +1,24 @@
+        this.playTime = playTime;
+    }
+
+    public void decrease() {
```
이전 리뷰에서 이부분 관련해서 값 객체라는 용어를 잘못 사용했네요. 🙇‍♂️
객체 이름으로 명시적으로 알 수 있다로 정정하겠습니다 :)

lalize commented on src/main/java/racingcar/model/PlayTime.java:
```diff
@@ -0,0 +1,24 @@
+    private static final int MIN_PLAY_TIME = 1;
+    private static final String MIN_PLAY_TIME_ERROR_MESSAGE = "[ERROR] 레이싱 횟수는 1회 이상이어야 합니다.";
+
+    private Integer playTime;
```
Wrapper 타입을 사용한 이유가 궁금합니다!

lalize commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,39 @@
+    private final Location location;
+    private final int movableNumber;
+
+    public Car(String carName) {
```
```suggestion
    public Car(String name) {
```

lalize commented on src/main/java/racingcar/model/Participants.java:
```diff
@@ -0,0 +1,54 @@
+        return this.cars.size();
+    }
+
+    public void race() {
+        for (RacingCar car : cars) {
+            car.tryMove(makeRandomNumberBetweenZeroAndNine());
+        }
+    }
```
[메서드 시그니처를 수정하여 테스트하기 좋은 메서드로 만들기](https://tecoble.techcourse.co.kr/post/2020-05-07-appropriate_method_for_test_by_parameter/)

lalize reviewed (approved):
안녕하세요 에덴!
피드백 반영 잘해주셨습니다 👍
추가 코멘트 남겨두었으니 확인하시고 다음 단계에서 진행 부탁드릴게요!
1단계는 여기서 머지하겠습니다~
